crates/models/llama/src/lib.rs

```diff
@@ -34,11 +34,22 @@ pub struct Llama {
 
     // must be kept alive for the model
     context: Arc<ggml::Context>,
+    loaded_tensors: HashMap<String, ggml::Tensor>,
 }
 
 unsafe impl Send for Llama {}
 unsafe impl Sync for Llama {}
 
+impl Drop for Llama {
+    fn drop(&mut self) {
+        for (_, tensor) in self.loaded_tensors.drain() {
```

Comment:
I'd suggest turning `loaded_tensors` into a struct that wraps the tensors, and does the transfer on `Drop`. That way, there's no risk of forgetting to do the transfer for each model.

An alternative suggestion would be to create a wrapper `OffloadableTensor(ggml::Tensor)`, which does the transfer back on `Drop`. That way, you can explicitly choose to opt into that behaviour for model tensors.

Reply:
I was thinking about moving this into the `ggml::context` as an optional `loaded_tensors` field, which is populated by the loader. As far as im aware  each model keeps its own `ggml::context` which means the context is dopped if the model is droppen. Meaning we just have to free the `loaded_tensors` when a context is dropped.

Reply:
Honestly it's a little surprising that the context doesn't clean this up itself when it's free.

I think having the context be mutated by the loader might be a bit weird in terms of cross-cutting concerns. Maybe have `Tensor::transfer_to` update the context's internal list of offloaded tensors (making sure to remove a tensor from that list if it's back on the CPU), and then having the context clean up that list on `Drop`?

Reply:
Solved it with some horrible `Arc<Mutex<HashMap>>` thingy. There probably is a better way of doing this 😓